lib/dotcom/system_status/groups.ex

```diff
@@ -0,0 +1,465 @@
+    end
+  end
+
+  # Exchanges a route_id (a line_id or a branch_id - anything that
+  # might correspond to an alert) for a map with that route_id and the
+  # statuses affecting that route.
+  defp add_statuses_for_route(route_id, alerts, time) do
+    %{
+      route_id: route_id,
+      statuses: statuses_for_route(route_id, alerts, time)
+    }
+  end
+
+  defp statuses_for_route(route_id, alerts, time) do
+    route_id
+    |> alerts_for_route(alerts)
+    |> alerts_to_statuses(time)
+  end
+
+  # Returns a branch_with_status entry, to be used in the
+  # branches_with_statuses field in groups/2. If no branch_ids are
+  # provided, then uses an empty array.
+  # 
+  defp branch_with_statuses_entry(statuses, branch_ids \\ []) do
+    %{
+      branch_ids: branch_ids,
+      statuses: statuses
+    }
+  end
+
+  # Given `alerts` and `route_id`, filters out only the alerts
+  # applicable to the given route, using the alert's "informed
+  # entities".
+  defp alerts_for_route(route_id, alerts) do
+    alerts
+    |> Enum.filter(fn %Alert{informed_entity: informed_entity} ->
+      informed_entity
+      |> Enum.any?(fn
+        %{route: ^route_id} -> true
+        %{} -> false
+      end)
+    end)
+  end
+
+  # Maps a list of alerts to a list of statuses that are formatted
+  # according to the system status specifications:
+  # - Identical alerts are grouped together and pluralized.
+  # - Times are given as a kitchen-formatted string, nil, or "Now".
+  # - Statuses are sorted alphabetically.
+  defp alerts_to_statuses(alerts, time) do
+    alerts
+    |> alerts_to_statuses_naive(time)
+    |> consolidate_duplicate_descriptions()
+    |> sort_statuses()
+    |> stringify_times()
+    |> maybe_add_now_text()
+  end
+
+  # Naively maps a list of alerts to a list of statuses, where a
+  # status is a simple structure with a route, a description, and a
+  # few additional fields that determine how it will render in the
+  # frontend.
+  defp alerts_to_statuses_naive(alerts, time)
+
+  # If there are no alerts, then we want a single status indicating
+  # "Normal Service".
+  defp alerts_to_statuses_naive([], _time) do
+    [%{description: "Normal Service", time: nil}]
+  end
+
+  # If there are alerts, then create a starting list of statuses that
+  # maps one-to-one with the alerts provided.
+  defp alerts_to_statuses_naive(alerts, time) do
+    alerts
+    |> Enum.map(fn alert ->
+      alert_to_status(alert, time)
+    end)
+  end
+
+  # Translates an alert to a status:
+  #  - The effect is humanized into a description for the status.
+  #  - If the alert's already active, `time` is set to `nil`.
+  #  - If the alert is in the future, `time` is set to the alert's
+  #    start time
+  defp alert_to_status(alert, time) do
+    description =
+      case alert.effect do
+        :delay -> "Delays"
+        :shuttle -> "Shuttle Buses"
+        :station_closure -> "Station Closure"
+        :suspension -> "Suspension"
+      end
```

Comment:
This could be a module attribute of a map or a function using pattern matching. What happens if the alert.effect isn't in this list?

Reply:
> This could be a module attribute of a map or a function using pattern matching.

True. This felt pretty much equivalent to both of those 🤷‍♂️.

> What happens if the alert.effect isn't in this list?

If the alert effect isn't in this list, that's an error that I'd probably want to see in Sentry. How do we typically handle this kind of thing?

Reply:
Usually pattern matching and the fallback logs the error with the argument.

Reply:
This code got eaten up by a different refactor, so it's gone now, and the `effect`'s are just passed through without being mapped to strings.